GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -76,17 +54,30 @@ public string ReadMessage()
                     throw new IOException("Incomplete message read from stream. The end of the stream was reached without the expected terminating byte.");
                 }
             }
+            while (readByte != TerminatorByte);
 
             return Encoding.UTF8.GetString(bytes.ToArray());
         }
 
         /// <summary>
-        /// Read the next chunk of bytes from the stream.
+        /// Read a byte from the stream.
         /// </summary>
-        /// <returns>The number of bytes read.</returns>
-        private int Read()
+        /// <param name="readByte">The byte read from the stream</param>
+        /// <returns>True if byte read, false if end of stream has been reached</returns>
+        private bool ReadByte(out byte readByte)
         {
-            return this.stream.Read(this.buffer, 0, this.buffer.Length);
+            int numBytesRead = this.stream.Read(this.buffer, 0, 1);
+
+            if (numBytesRead > 0)

```

Comment:
I think you can drop this condition and always execute line 73. The caller won't use the value if we return false, so there's no point in doing the check twice.